src/Microsoft.AspNetCore.Mvc.DataAnnotations/Internal/DataAnnotationsMetadataProvider.cs

```diff
@@ -176,15 +176,22 @@ public void CreateDisplayMetadata(DisplayMetadataProviderContext context)
 
                 // EnumDisplayNamesAndValues and EnumNamesAndValues
                 //
-                // Order EnumDisplayNamesAndValues to match Enum.GetNames(). That method orders by absolute value,
-                // then its behavior is undefined (but hopefully stable). Add to EnumNamesAndValues in same order but
-                // Dictionary does not guarantee order will be preserved.
+                // Order EnumDisplayNamesAndValues by DisplayAttribute.Order, then by the order of Enum.GetNames(). 
+                // That method orders by absolute value, then its behavior is undefined (but hopefully stable). 
+                // Add to EnumNamesAndValues in same order but Dictionary does not guarantee order will be preserved.
+
                 var groupedDisplayNamesAndValues = new List<KeyValuePair<EnumGroupAndName, string>>();
                 var namesAndValues = new Dictionary<string, string>();
                 var enumLocalizer = _stringLocalizerFactory?.Create(underlyingType);
-                foreach (var name in Enum.GetNames(underlyingType))
+                var enumNames = Enum.GetNames(underlyingType);
+
+                var enumFields = from i in Enumerable.Range(0, enumNames.Length)
+                                 let field = underlyingType.GetField(enumNames[i])
+                                 orderby field.GetCustomAttribute<DisplayAttribute>(inherit: false)?.GetOrder(), i
```

Comment:
Should be explicit about the where `enum` values without a `[Display]` attribute or `Order` setting appear. `DisplayAttribute.Order` [documentation](https://msdn.microsoft.com/en-us/library/system.componentmodel.dataannotations.displayattribute.order(v=vs.110).aspx) mentions "presentation layers should consider using the value 10000." `DisplayAttribute.GetOrder()` [documentation](https://msdn.microsoft.com/en-us/library/system.componentmodel.dataannotations.displayattribute.getorder(v=vs.110).aspx) says the same thing. However, we tend to use `1000` (an order of magnitude less) elsewhere in MVC. My suggestion is therefore
``` c#
.OrderBy(field => field.GetCustomAttribute<DisplayAttribute>(inherit: false)?.GetOrder ?? 1000)
```

The `ThenBy()` part included here shouldn't be needed because `System.Linq` will preserve order.

More generally, I'm concerned about duplicates. Please add a test demonstrating that _all_ names in the following are included in the two collections. Feel free to add some `Order` values of course 😈 
``` c#
enum Joe
{
  Fred = 0,
  Ginger = 0,
  Larry = 1,
  Curly = 1,
  Moe = 1,
}
```